Arduino-mega2560 GPIO implementation

ReneHerthel opened the pull request:
-Implements some GPIO functionality for arduino-mega2560, with cpu ATmega2560, for pins 2 -13  at PWM

- Add missing functions: gpio_init_out / init_in / read / set / clear / toggle / write.

PeterKietzmann commented:
@ReneHerthel nice! Will review soon

PeterKietzmann commented on boards/arduino-mega2560/include/periph_conf.h:
```diff
@@ -264,23 +265,67 @@ extern "C" {
-#define GPIO_14_EN          0
-#define GPIO_15_EN          0
+#define GPIO_NUMOF          12
+#define GPIO_2_EN           1
```
Why didn't you implement `GPIO_0` and `GPIO_1`?

ReneHerthel replied:
It's because there are pin 2-13 at PWM- and 0-1 in communication-section on the pinout.
for example: if you want to set pin 2, you have to call the method with 2 (parameter: dev).
But when you change the pin_map, that GPIO_0 is first, you will have the wrong pin all time, so i decided to implement it this way.

ReneHerthel replied:
So, should i change it from GPIO_2 to GPIO_0... ?

PeterKietzmann replied:
The advantage of the periph_conf is that you can abstract the hardware-dependent stuff from your driver. That means the "GPIO_0" could for example either be port E pin 5 or port H pin 6... Please change this and all dependencies. You don't need to implement 15 GPIO pins (even if it would be nice :-) ). It is more important to prevent potential pin conflicts for later driver implementations. Tha means: Try to avoid implemention GPIOs when blocking other peripheral pins with that. Example: If there are just 3 pins that can be configured as SPI-pins, don't implement the GPIO functionality on them. Ok?

PeterKietzmann commented:
I don't know this board in detail. Where is the relation between PWM and GPIO? Isn't it possible to implement them independently?

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -28,22 +27,132 @@
 {
-    return -1;
+    uint8_t pin;
+    pin = gpio_pin_map[dev - 2];
```
I know, you started your implementation at `GPIO_2` but I don't know the reason for this. What if one tries to initialize `GPIO_0 ` (which has the value `0`) here?

ReneHerthel replied:
Didn't noticed that. Thanks

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -28,22 +27,132 @@
+    pin = gpio_pin_map[dev - 2];
+    switch(dev) {
+      case 2:
+        DDRE  |= (1 << pin);
```
What is the `DDRE`? Is it related to the port? You never use your port definitions from the periph_conf.h. Could it be reasonable to blend these makros somehow?

ReneHerthel replied:
It's the Data Direction Register. You can define there which direction the pin is used for.
Yes it is. DDR "E" is  related to PORT "E" and so on.
The makros belongs to avr/io.h, i thought that i could use it this way.
I can put it in the periph_conf.h file.

PeterKietzmann replied:
Comparing other CPUs these registers are defined as structs for each port. One has to look in the avr/io.h in detail. If this is not the case, I'd vote for definig this in the periph_conf.h. Could like alike:
```
#define GPIO_2_PORT         PORTE
#define GPIO_2_REG_DIR      DDRE
#define GPIO_2_PIN          PE4

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -28,22 +27,132 @@
+        DDRB  |= (1 << pin);
+        break;
+    };
+    return 0; /* Everything alright */
```
We need to implement the pushpull setup in future (can start with this together)

ReneHerthel replied:
Okey!

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -28,22 +27,132 @@
+        default:
+          return -1;
+    };
+    return 0; /* Everything alright */
```
We need to implement the pushpull setup in future (can start with this together)

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -28,22 +27,132 @@
 {
-    return -1;
+    uint8_t pin;
+    pin = gpio_pin_map[dev - 2];
```
same as above

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -53,23 +162,137 @@ int gpio_init_int(gpio_t dev, gpio_pp_t pushpull, gpio_flank_t flank, gpio_cb_t
 {
-    return -1;
+    uint8_t pin;
+    pin = gpio_pin_map[dev - 2];
```
same as above

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -53,23 +162,137 @@ int gpio_init_int(gpio_t dev, gpio_pp_t pushpull, gpio_flank_t flank, gpio_cb_t
+      default:
+        return -1;
+    };
+    return 0; /* Everything alright */
```
Looking at the gpio interface in drivers/include/periph/gpio.h the `gpio_read()` function should return a `0` when the input pin is low and a `>0` when the pin is high

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -53,23 +162,137 @@ int gpio_init_int(gpio_t dev, gpio_pp_t pushpull, gpio_flank_t flank, gpio_cb_t
 void gpio_set(gpio_t dev)
 {
+    uint8_t pin;
+    pin = gpio_pin_map[dev - 2];
```
same as above

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -53,23 +162,137 @@ int gpio_init_int(gpio_t dev, gpio_pp_t pushpull, gpio_flank_t flank, gpio_cb_t
+    pin = gpio_pin_map[dev - 2];
+    switch(dev) {
+      case 2:
+        PORTE |= (1 << pin);
```
Why don't you use the defined port macro from the periph_conf.h?

ReneHerthel replied:
I tried to implement a port_map like the pin_map, but i failed, because the compiler said, the initializer of GPIO_0...GPIO_1 ... is not constant.
I think i must find a different way to implement that.

PeterKietzmann replied:
Can you give a rudimentary example of what you did? Otherwise: I meant you should use `GPIO_2_PORT` instead of `PORTE`

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -53,23 +162,137 @@ int gpio_init_int(gpio_t dev, gpio_pp_t pushpull, gpio_flank_t flank, gpio_cb_t
+void gpio_clear(gpio_t dev)
 {
+    uint8_t pin;
+    pin = gpio_pin_map[dev - 2];
```
same as above

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -53,23 +162,137 @@ int gpio_init_int(gpio_t dev, gpio_pp_t pushpull, gpio_flank_t flank, gpio_cb_t
+    pin = gpio_pin_map[dev - 2];
+    switch(dev) {
+      case 2:
+        PORTE &= ~(1 << pin);
```
same as above

PeterKietzmann commented:
The interrupt functionality is missing

PeterKietzmann commented:
@ReneHerthel nice work and congrats to your first RIOT-PR :-) !

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -28,22 +27,132 @@
+    uint8_t pin;
+    pin = gpio_pin_map[dev - 2];
+    switch(dev) {
+      case 2:
```
It would be much prettier using the GPIO names here. (So `2` will become `GPIO_2` and so on)

PeterKietzmann commented on boards/arduino-mega2560/include/periph_conf.h:
```diff
@@ -264,23 +265,110 @@ extern "C" {
+/* GPIO channel 1 configuration */
+#define GPIO_1_PORT         0 /* Not implemented yet */
+#define GPIO_1_PIN          0 /* Not implemented yet */
+#define GPIO_1_DDR          0 /* Not implemented yet */
```
Why don't you just configure them? You can for example just implement 12 or less GPIOs and move the pins from `GPIO_13` and `GPIO_14` to `GPIO_0` and `GPIO_1`. Or shift all up, or mix them up randomly :-)

PeterKietzmann commented on boards/arduino-mega2560/include/periph_conf.h:
```diff
@@ -264,23 +265,110 @@ extern "C" {
+#define GPIO_PIN_B          PINB
+#define GPIO_PIN_E          PINE
+#define GPIO_PIN_G          PING
+#define GPIO_PIN_H          PINH
```
Why is this needed? Well it doesn't hurn so much but there is no reason for it, right? In my opinion it would be totally fine to use `#define GPIO_2_PORT         PORTE` in the lower lines

PeterKietzmann commented on boards/arduino-mega2560/include/periph_conf.h:
```diff
@@ -264,23 +265,110 @@ extern "C" {
+
+/* GPIO channel 2 configuration */
+#define GPIO_2_PORT         GPIO_PORT_E
+#define GPIO_2_PIN          PE4  /* 0x10 0b00010000 PWM: Pin 2 */
```
I find it a bit irritating reading about PWM in the documentation for GPIO. Is this somehow reasonable?

PeterKietzmann commented on boards/arduino-mega2560/include/periph_conf.h:
```diff
@@ -264,23 +265,110 @@ extern "C" {
+/* GPIO channel 2 configuration */
+#define GPIO_2_PORT         GPIO_PORT_E
+#define GPIO_2_PIN          PE4  /* 0x10 0b00010000 PWM: Pin 2 */
+#define GPIO_2_DDR          GPIO_DDR_E
```
Same as above: `#define GPIO_2_DDR    DDRE` should be enough, right?

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -28,48 +27,500 @@
-    return -1;
+    uint8_t pin;
+
+    if (dev > GPIO_NUMOF || dev < 2)
```
Still I am of that opinion to just start the implementations at `GPIO_0`. Then you don't neet to check for `dev < 2`. Also the API says `return -1` means "pull-up mode unavailable". But in general the devive available-check is a good thing

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -28,48 +27,500 @@
+
+    switch (dev)
+    {
+        case 0:
```
Would be nice and better readable if you substitute `case 0` by `case GPIO_0` and following

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -28,48 +27,500 @@
+        case 1:
+            return -1;
+          /*GPIO_1_DDR |= (1 << pin);
+            break;*/
```
As alreaedy said, just implement the first two GPIOs

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -28,48 +27,500 @@
-    return -1;
+    uint8_t pin;
+
+    if (dev > GPIO_NUMOF || dev < 2)
```
same as above

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -28,48 +27,500 @@
+        case 1:
+            return -1;
+          /*GPIO_1_DDR &= ~(1 << pin);
+            break;*/
```
same as above

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -28,48 +27,500 @@
+
+    switch (dev)
+    {
+        case 0:
```
same as above

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -28,48 +27,500 @@
+ * when bit == 0,
+ *  -return 0 (LOW),
+ *  -else return 1 (HIGH)
+ */
```
??? The documentation of the API is places in drivers/include/periph/gpio.h

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -28,48 +27,500 @@
-    return -1;
+    uint8_t pin;
+
+    if (dev > GPIO_NUMOF || dev < 2)
```
I'll stop commenting this here, same as above..

PeterKietzmann commented:
@ReneHerthel sorry for the vast number of comments. This is normal at the beginning. If you see problems or don't know how to handle anything, just ask here (best way), send me a private mail or come around here :-)

PeterKietzmann commented on boards/arduino-mega2560/include/periph_conf.h:
```diff
@@ -264,23 +265,79 @@ extern "C" {
+/* GPIO channel 0 configuration */
+#define GPIO_0_PORT         PORTE
+#define GPIO_0_PIN          PE4
+#define GPIO_0_DDR          DDRE
```
That is wat I meant :-)

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -28,24 +27,167 @@
 
-/*static gpio_state_t config[GPIO_NUMOF]; */
-
+static gpio_state_t config[GPIO_NUMOF];
```
What do you use it for?

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -53,23 +195,262 @@ int gpio_init_int(gpio_t dev, gpio_pp_t pushpull, gpio_flank_t flank, gpio_cb_t
+            {
+                return 0;
+            }
+            break;
```
If you introduce a port map you could shrink this whole "sausage" :-) ! Compare the stm32f4 implementation [here](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32f4/periph/gpio.c#L44) and [here](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32f4/periph/gpio.c#L265). Btw: The same applies for the `DDR` but it's up to you

PeterKietzmann replied:
Ah and also have a look [here](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32f4/periph/gpio.c#L467)

ReneHerthel replied:
Yes, i would like to implement it this way, but i don't know how, because the PORTx seems not to be constant and the initializer element in the port map has to be.
hmm.. i tried it already before..

ReneHerthel replied:
How does the struct 'GPIO_TypeDef' works? ... first link.

PeterKietzmann replied:
Afaik GPIO_TypeDef is specific to the ARM family and therefore shouldn't be your reference. I'll try to test this by myself soon and give feedback on it then.

PeterKietzmann commented:
@ReneHerthel everything look much better! Some hints and questions:
1. Please trigger the Assignee of a PR when you've committed new stuff. Something like "@PeterKietzmann I addressed your comments" is fine
2. Are you planning to follow up the work on this driver? I'm thinkink especially about the interrupt implementation
3. At any time in point I need to borrow your board to test the implementation (if you want me to merge this PR ;-) )

ReneHerthel commented:
1. How can i trigger you? ... by creating a new comment and write @PeterKietzmann.. ?
2. Yes.
3. It's okey, i've two arduino's.

PeterKietzmann commented:
1. Yes, just write a comment 
2. Nice! So assume you will publish your results in this PR
3. Also nice! Maybe you can just come around during the next days and let one there

PeterKietzmann commented:
Please also add an entry in the `Makefile.features` like [here](https://github.com/RIOT-OS/RIOT/blob/master/boards/stm32f4discovery/Makefile.features#L3)

PeterKietzmann commented:
If you look at the avr files (compare *your_install_path*/avr/include/avr/iomxx0_1.h and  *your_install_path*/avr/include/avr/srf_defs.h) I think this is the offender. The `PORTE` calls a macro which computes a hex-address. I don't know how to solve this issue out of the box so I'd vote to begin with other tasks first

ReneHerthel commented:
@PeterKietzmann i've added the feature in the makefile. 

Okey then.. let us start with the pushpull or interrupt implementation?

PeterKietzmann commented:
Öhm, okey! Do you just want to start? Or do you need more help? As a quick introduction you can read [this](http://coactionos.com/embedded%20design%20tips/2013/10/21/Tips-Understanding-Microcontroller-Pin-Input-Output-Modes/) summarization

ReneHerthel commented:
@PeterKietzmann The pull-up resistor will be enabled simultaneously, when a bit (for a certain pin) is set in the PORTx register.So i added some code for the pushpull in gpio_init_out / in.
I've also added a new register, the PINx register, which represent the state of a certain bit.

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -28,24 +27,232 @@
+    {
+        case GPIO_0:
+            GPIO_0_DDR |= (1 << pin);
+            if (pushpull)
```
@ReneHerthel look at the possible [pushpull modes](https://github.com/RIOT-OS/RIOT/blob/master/drivers/include/periph/gpio.h#L142) in RIOT. With your implementation you will set the pull-ups even if the `GPIO_PULLDOWN` is wanted. What I didn't know was that the MCU only supports "no pull-ups" or "pull-ups" and no "pull-downs". Comparin the gpio [API](https://github.com/RIOT-OS/RIOT/blob/master/drivers/include/periph/gpio.h#L167) you can see that you have to `return -1` on an unavailable mode. Compare the implementation for the [arduino-due CPU](https://github.com/RIOT-OS/RIOT/blob/master/cpu/sam3x8e/periph/gpio.c#L498)

PeterKietzmann replied:
... here and all following

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -53,23 +260,188 @@ int gpio_init_int(gpio_t dev, gpio_pp_t pushpull, gpio_flank_t flank, gpio_cb_t
+    switch (dev)
+    {
+        case GPIO_0:
+            if ( GPIO_0_STATE & (1 << pin) )
```
I assume the only "possible" values you can read from this registers will be `0` and `1`. Why don't you just `return` the read value? You will save many lines of `if else`

PeterKietzmann commented on boards/arduino-mega2560/include/periph_conf.h:
```diff
@@ -264,23 +265,56 @@ extern "C" {
+#define GPIO_0_PORT         PORTA
+#define GPIO_0_PIN          PA0
+#define GPIO_0_DDR          DDRA
+#define GPIO_0_STATE        PINA
```
I don't like the name of `GPIO_0_STATE ` but neither I have a better idea atm. It's just because the other defines are somehow related to the actual register names but "state" is not. Hm..

PeterKietzmann commented:
@ReneHerthel nice work and thanks for your patience :-) ! BTW did you read the [datasheet](http://www.atmel.com/Images/Atmel-2549-8-bit-AVR-Microcontroller-ATmega640-1280-1281-2560-2561_datasheet.pdf) chap. 13 starting on page 67?

ReneHerthel commented:
@PeterKietzmann So I changed some useful thinks I guess:
- Some macros in periph_config.h, particularly names.
- Reduced the code, particularly all if-else statements.

By the way, what does "NEEDS SQUASHING" mean? :-)

PeterKietzmann commented:
@ReneHerthel I will have a look at yout changes during the Hack'n'ACK today. BTW: Will you be there too? "NEEDS_SQUASHING" is a reminder to not forget squashing all commits to one single commit with a reasonable naming before merging the PR.  Please read our [Developement Procedures](https://github.com/RIOT-OS/RIOT/wiki/Development-procedures). You can find some informations about realising a squash in our [git cheatsheet](https://github.com/RIOT-OS/RIOT/wiki/Git-cheatsheet). Please note that (in general) you just squash after the PR is ACKed though the assignee can see the diffs of a commit that addresses comments. Understand :D ? . In your case I guess you can squash all commits except the last two ones.

PeterKietzmann commented:
@ReneHerthel which test did you use to verify the functionality? Could you point to a place in your repo, post the snipped here oder add a commit which you remove just before merging? It was not easily testable for me. Seems timers and UART are somehow insufficient for this board

ReneHerthel commented:
@PeterKietzmann I've written my own simple main.c file. But I tested it just now with RIOT/test/periph_gpio/main.c and it works also. You mean the console-output? Yes.. there are many repetitions.. but I don't know if it's normal or not with other boards.

PeterKietzmann commented:
Well, this is not normal :-) ! Can you please provide your main.c so I don't need to write an own or adapt the tests/periph_gpio. (Btw I guess this strange behavior is not related to your driver)

ReneHerthel commented:
@PeterKietzmann So i changed the names of the registers and put the pullup retrieval in a switch-case. I've just written a simple file, with an for-loop.. but i can send it to you via email.
Now i want to start with the interrupt thing, but only if there is nothing more to discuss :-)

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -28,22 +27,169 @@
+            break;
+        case GPIO_PULLDOWN:
+            return -1;
+    };
```
Did you overread my comment: "*I thought about it a bit. Let's just leave the pull-up configuration out for the first. It seems to be handled by hardware in a "senseful" way. Just add a comment for this here. Sorry for the confusion*" ?
As we read in the manual the pull-up handling is half automatic and not separately controllable for each pin. So let's just take it out.

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -53,23 +199,175 @@ int gpio_init_int(gpio_t dev, gpio_pp_t pushpull, gpio_flank_t flank, gpio_cb_t
+    {
+        gpio_set(dev);
+    }
 }
```
This looks valid. Alternatively the manual saye "*However, writing a logic one to a bit in the PINx Reg-ister, will result in a toggle in the corresponding bit in the Data Register.*. Cou could also try that but don't spend too much time in this

PeterKietzmann replied:
Did you test this once?

PeterKietzmann commented:
@ReneHerthel everything looks MUCH better now! Please address my (small) comments and implement the interrupt stuff. Then we can test the stuff together

ReneHerthel commented:
@PeterKietzmann So I deleted the pushpull thing. Now I try to understand how the interrupts works.
I wrote a little bit.. maybe you have a clue?

OlegHahm commented:
Any news?

ReneHerthel commented:
@PeterKietzmann need a bit help I guess..
I found out how to use the registers, but I need more help to understand how the falling and rising edge should be implemented by software. 
Can you maybe explain me how the gpio_init_int(..) function works, with all parameters?
I can compile that, but the interrupt only triggers in a strange way..

PeterKietzmann commented:
@ReneHerthel yes. Let's look at this in the next week. This week is a bit stressy for me..

ReneHerthel commented:
@PeterKietzmann So any ideas?

ReneHerthel commented:
@PeterKietzmann I couldn't find a way to disable the interrupt source by writing a logic 1 into the register. It doesn't work, because the board do this automatically.

OlegHahm commented:
Btw. your local gitconfig seems to be configured to a different e-mail address than the one(s) on Github, @ReneHerthel.

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -29,47 +31,440 @@
+        case GPIO_5:
+            EIMSK |= (1 << pin);
+            break;
+#endif /* GPIO_5_EN */
```
Please also implement the API functions `gpio_irq_enable()` and `gpio_irq_disable()`. If you like, you could then (not must) also call the enabling function with the regarding `dev` here.

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -29,47 +31,440 @@
+#endif /* GPIO_5_EN */
+    };
+
+    pin_B = (pin > GPIO_3 && pin == GPIO_4) ? 0 : 1;
```
I don't now exactly what you are doing here. Are you sure the `&&` is correct? Anyway, do you know [modulo operations](http://en.wikipedia.org/wiki/Modulo_operation)? Let's do it like this, without using the above line :
```
    /* configure the flank */
    switch (flank) {
        case GPIO_RISING:
            if (pin < 4)
            {
                EICRA |= (3 << pin * 2);
            }
            else
            {
                EICRB |= (3 << (pin * 2)%4));
            }
            break;
```

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -29,47 +31,440 @@
+    if (sched_context_switch_request) {
+        thread_yield();
+    }
+    */
```
Why is this commented out?

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -29,47 +31,440 @@
+}
+
+#if GPIO_0_EN
+ISR(INT0_vect, ISR_BLOCK)
```
I think you also need to abstract `INT0_vect` in the periph_conf file. Imagine The pin of `GPIO_0` changes. Would this still be the correct interrupt? I don't know exactly. So don't change just in cause of this comment :-) !

PeterKietzmann commented:
@ReneHerthel I'm happy to see that the driver basically works know, nice work! There are just some coding style issues left. I will test and hopefully merge when my comments were addressed :-) .

PeterKietzmann commented:
Just has a reminder: We should be aware of #3095

ReneHerthel commented:
Uhh, okey! I will have a look on it, next week.

ReneHerthel commented:
@PeterKietzmann What's up with the last periph driver changes?

[3095 changed periphal Drivers] (https://github.com/RIOT-OS/RIOT/pull/3095)

I think i should finish this in my PR right?

PeterKietzmann commented:
@ReneHerthel it would be really cool if you could finish this implementation, yes :-) ! In #3426 we figured out that the shell stuff works well if you set the defualt baudrate for UART to 9600. 
All changes to the gpio drivers had been described (and adapted) in #3095. What exactly is the question? To avoid unrelated problems you should quit using timers in you personal test applications.

ReneHerthel commented:
@PeterKietzmann No questions at the moment, just want some informations like this, thanks :-) .. I'll work on it then.

ReneHerthel commented:
Rebase branch.

ReneHerthel commented:
@PeterKietzmann Is it possible that I have to implement the new GPIO cpu definitions in the file: [cpu\atmega2560\periph_cpu.h](https://github.com/RIOT-OS/RIOT/blob/master/cpu/atmega2560/include/periph_cpu.h) too?
Because the other boards have there some definitions about the GPIO, and there is just a comment says 'nothing defined here so far...'.

PeterKietzmann commented:
@ReneHerthel yes, this is possible :-) . But I won't say "too" but more "instead". We can have a look at it together, if you see problems. 
Moreover it looks like something went wrong during your rebase. Did you see our guide [git cheatsheet](https://github.com/RIOT-OS/RIOT/wiki/Git-cheatsheet) sec. *How to rebase a branch on the current RIOT master*? 
As this PR was really old, you can also squash all your commits to one single.

ReneHerthel commented:
@PeterKietzmann second rebase, it's now ok?

PeterKietzmann commented:
Ja, looks way better. So are you planning to adapt the driver?

ReneHerthel commented:
@PeterKietzmann Yes :-) .. I really want to finish this PR!

ReneHerthel commented:
@PeterKietzmann Okey I have many questions I think.
On the [#3095](https://github.com/RIOT-OS/RIOT/pull/3095) are many informations how to adapt that, but also many changes during the PR-process. So the first comment didn't introduce well how to do it.
I'm looking at the stm boards, but they have some different implementation like the 'GPIO_TypeDef' and so on..
So the main question is how to get the pin number and port number?
AND what does the periph_cpu.h file do? 

especially:

```Javascript
#define HAVE_GPIO_T
typedef uint32_t gpio_t;

...

#define GPIO(x, y)     ((GPIOA_BASE + (x << 10)) | y)

... 

#define GPIO_UNDEF    (0xffffffff)
```
link: [stm32f4 -> periph_cpu.h](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32f4/include/periph_cpu.h).

Are there other Things I  should mind?

PeterKietzmann commented:
@ReneHerthel I also need to think about it. But also the API changed slightly. You could adapt that in in first step.

PeterKietzmann commented:
@ReneHerthel, the main idea is to have a macro (`GPIO(x, y)`) that handles port and pin informations. Additionally, some helper functions in the driver implementation can return the regarding pin number, port number or the real port address. This gives you, in general, a higher flexibility regarding pin configurations and removes the need of all this preprocessor commands.  
I see, that the 32bit arm platforms are bit different, therefore you can not crib so much ;-) . 
If you have a look at the register overview of the [atmega2560](http://www.atmel.com/Images/Atmel-2549-8-bit-AVR-Microcontroller-ATmega640-1280-1281-2560-2561_datasheet.pdf) (sec. 33, p. 399) you can find some similarities for each port. I'd suggest to have these (helper-) functions in the driver: 
```
/* Returns number of the port. E.g. PortA: 0, PortB: 1, PortC: 2 ... */
uint8_t  _port_num(gpio_t pin);
/* Returns pin number. E.g. PA0: 0, PB4: 4, PC6: 6 ... */
uint8_t  _pin_num(gpio_t pin);
/* Returns the PORT register address. E.g. PortD: 0x2B, PortA: 0x22 ... */
uint8_t *_port_addr(gpio_t pin);
/* Returns the DDR register address. E.g. DDRC: 0x27, DDRB: 0x24 ... */
uint8_t *_ddr_addr(gpio_t pin);
```
You need to find a trick how to safe the combination of port, pin and address and how to extract these informations again. 
Understand the main idea? Just PM me or come around if you need more help. But I guess you will make it somehow!

PeterKietzmann commented:
@ReneHerthel you can have a look at this PR #3535 to see the effect of this rework

ReneHerthel commented:
@PeterKietzmann Yes I understand, the only thing is this macro in [gpio.h](https://github.com/RIOT-OS/RIOT/blob/master/drivers/include/periph/gpio.h#L34):
```Javascript
/**
 * @brief   Default GPIO macro maps port-pin tuples to the pin value
 */
#ifndef GPIO
#define GPIO(x,y)       ((x & 0) | y)
#endif
```
Can you explain how to use it?

In this file [stm32f4->periph_cpu.h](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32f4/include/periph_cpu.h#L45) there is this macro in use, but I dont understand  what it does.

It seems that this is an important Point.

kaspar030 commented:
> Can you explain how to use it?

That's a convenience macro so someone can write ```gpio_set(GPIO(PORT_A,
5)``` instead of ```gpio_set(0x12340005)``` or whatever encoding the
implementation is using.

ReneHerthel commented:
Aaaah okey, thanks :-)

PeterKietzmann commented:
@ReneHerthel in addition: In stm32f4/periph_cpu.h the macro is not used but defined for the stm32f4 CPU. In gpio.h it is just a "default" definition **if** `GPIO` (the macro's name) is not defined elsewhere.

PeterKietzmann commented:
In an application the Usage could look like [this](https://github.com/RIOT-OS/RIOT/blob/master/tests/periph_gpio/main.c#L74)

ReneHerthel commented:
@PeterKietzmann Okey, and with These helper-functions we can extract the addres, port/pin number and ddr, right?

I'll push my work son, then you can have a look :)

ReneHerthel commented:
@PeterKietzmann So the next Problem:
We accessed the ports over the macros like:
```Javascript
                        ...

GPIO_2_DDR_REG &= ~(1 << pin);

                        ...
```
In the old implementation..
Now, I have something like that:
```Javascript
                        ...

_ddr_addr(pin) &= ~(1 << _pin_num(pin));

                        ...
```

I think that it's not work properly.
Because I dont know how to use These avr/io.h macros with the new implementation..

PeterKietzmann commented:
> I'll push my work son, then you can have a look :)

I will have a closer look tomorrow, "father" ;-) ! 
The general idea you described is reasonable, happy to see you got what I tried to explain! Ideally you won't need these avr/io.h defines later. What stood behind `GPIO_2_DDR_REG` was just an address that you will get out of `_ddr_addr(pin) ` now. 
Now looking at the stm32f4 as you tried to compare it. `GPIOA_BASE` is a register address, where the the IO related registers start. As `A` is the first "bank", it is used as "base". Now look at `((GPIOA_BASE + (x << 10)) | y)` with `x`: port as number and `y`: pin number. If `x=1` and `y=0` you will return the `GPIOB_BASE` address and if `x=2` and `y=0` you will return the `GPIOC_BASE` address...  If `x=0` and `y=7` you will return the `GPIOA_BASE + "pin7"` address. ` (x << 10)` in general does the trick to increment through the base addresses with incrementing `x`. `y` it written in the lower bits to put the regarding pin number in that value. The drivers macros (e.g. `_pin_number(pin))`) will extract this information again. Got what I mean? It helped me a lot to write down the binary values and compare the regarding binary addresses out of the CMSIS header.
Now back to the atmega: In general I thought you could also use `PORTA` as a base address and find a trick to access the ports by incrementing an `x`. I've had no thoughts about the pin number until now.

ReneHerthel commented:
@PeterKietzmann Hm I got several issues, when I try to run the gpio test application..
But it seems that the problem is located to the gpio.h file?

```Javascript
root@rene-ubuntu:~/Repositorys/RIOT/tests/periph_gpio# BOARD=arduino-mega2560 make
Building application "periph_gpio" for "arduino-mega2560" with MCU "atmega2560".

In file included from /home/rene/Repositorys/RIOT/tests/periph_gpio/main.c:27:0:
/home/rene/Repositorys/RIOT/drivers/include/periph/gpio.h:55:9: error: expected ‘;’, identifier or ‘(’ before ‘enum’
 typedef enum {
         ^
/home/rene/Repositorys/RIOT/drivers/include/periph/gpio.h:100:27: error: expected declaration specifiers or ‘...’ before ‘gpio_dir_t’
 int gpio_init(gpio_t pin, gpio_dir_t dir, gpio_pp_t pullup);
                           ^
/home/rene/Repositorys/RIOT/tests/periph_gpio/main.c: In function ‘init_out’:
/home/rene/Repositorys/RIOT/tests/periph_gpio/main.c:74:5: error: implicit declaration of function ‘gpio_init’ [-Werror=implicit-function-declaration]
     if (gpio_init(GPIO(port, pin), GPIO_DIR_OUT, pull) < 0) {
     ^
cc1: all warnings being treated as errors
make[1]: *** [/home/rene/Repositorys/RIOT/tests/periph_gpio/bin/arduino-mega2560/periph_gpio/main.o] Fehler 1
make: *** [all] Fehler 2
root@rene-ubuntu:~/Repositorys/RIOT/tests/periph_gpio#
```

Did I missed something?
Is it new that all warnings being treated as errors?

I think I should be able to compile this test and this branch is up-to-date.

PeterKietzmann commented on cpu/atmega2560/include/periph_cpu.h:
```diff
@@ -20,12 +20,55 @@
+    PIN_5 = 0x05,   /**< pin 5 */
+    PIN_6 = 0x06,   /**< pin 6 */
+    PIN_7 = 0x07    /**< pin 7 */
+}
```
You forgot the semicolon here

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -0,0 +1,194 @@
+        gpio_clear(pin);
+    }
+}
+
```
I think you can remove the below lines.

PeterKietzmann commented:
@ReneHerthel I didn't completely understand the stuff in periph_cpu.h on the first glance. Therefore I'd like to see your outcomes from the compilation and the test at first and further discuss then.

ReneHerthel commented:
@PeterKietzmann I fixed some bugs and can compile it.
Now I am able to extract the pin and port number, extracting the port-/ddr address doesn't work properly. 

I wrote the pin number into the last 4 bits and the port number into the first 4 bits.

There's also a special case, where the first 7 ports (A-G) are on address 0x22 - 0x31, the other 4 (H - L) are on address 0x102 - 0x10B.
The ddr addresses are the same as the correspending port address but subtracted by one.

The problem is maybe [these](https://github.com/RIOT-OS/RIOT/commit/35c14f962884b22fc894d095b1407165967fd943#diff-5a58a4ac457ecb836c666ca8d4f62627R64) line.

I'll have a look after next week.

PeterKietzmann commented:
@ReneHerthel I sent you a private mail with example code. The trick is in _avr/include/avr/srf_defs.h:176 and 128_ . Please ping me if you got it working *or* need further help. I think we are on the right track now :-)

ReneHerthel commented:
@PeterKietzmann That was very helpful, thanks! Now it should work :-).

The next thing is the interrupt implementation.

PeterKietzmann commented:
@ReneHerthel nice to hear that! Also it seems we really save a lot of code with this implementation, which is also a success! Unfortunately I won't find time to test your implementation before the (end?! of) next week. Do you need further help regarding the interrupts? 
BTW: Please don't comment con single commits.  What did you want to tell me with that link above?

kaspar030 commented on cpu/atmega2560/include/periph_cpu.h:
```diff
@@ -20,12 +20,33 @@
+    PORT_E = 4,       /**< port E */
+    PORT_F = 5,       /**< port F */
+    PORT_G = 6,       /**< port G */
+    PORT_H = 0,       /**< port H */
```
Not looking through the rest of the code, why are ports sharing an id?

PeterKietzmann replied:
I neither looked at the code nor checked details, but just as an assumption: Won't it be enough to just support up to PORT_F? IIRC there was a shift in the registers behind PORT_F or PORT_G and these pins aren’t available on the arduino-mega2560 anyway. (We're on the CPU specific and not board specific part here, I know I know :-) )

ReneHerthel replied:
@kaspar030 Because there is a gap between the base address of PORT_G and PORT_H (about 0xE0), the addresses of the ports are not continously. Later the address will be generated in gpio.c and I have to differentiate between these Ports. So I implemented two base addresses, one from the base of A (0x20) and one from the base of H (0x100). Therefore I Count at Port_H from 0 again and can easily generate the port-address.

@PeterKietzmann Okey, that can be a solution, if you want :-) Should I Change that?

kaspar030 replied:
> Therefore I Count at Port_H from 0 again and can easily generate the port-address.

But that means GPIO(PORT_A, 0) and GPIO(PORT_H, 0) are equivalent,
right? Or are, e.g., port H's pin numbers in a different number space
than port A's?

ReneHerthel replied:
@kaspar030 Hmm yes, I just used this for the calculation of the port address. But I found a little bug, so I changed it back to PORT_H = 7, PORT_J = 8 and so on.
Now I'm using an Offset to Port_H instead of a Base address, when the Port number is higher then PORT_G. So all pins are accessible and the 'ugly' thing in the periph_cpu.h is gone :-)

kaspar030 commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -0,0 +1,215 @@
+
+    port_num = _port_num(pin);
+
+    if (port_num < PORT_H) {
```
together with above enum, please clarify. ;)

ReneHerthel commented:
@PeterKietzmann Just want to link the other macro, if you want to check this.. :-P, sorry.
Hmm, maybe later I do have some questions :-)

ReneHerthel commented:
@PeterKietzmann Adapt the interrupt functionality.

PeterKietzmann commented:
@ReneHerthel thanks for your further work. Running tets/periph_gpio I don't get the correct values with `init_in... ; read...;` . Any ideas? did that work for you?

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -0,0 +1,244 @@
+    return 0;
+}
+
+int gpio_init_int(gpio_t pin, gpio_pp_t pullup, gpio_flank_t flank,
```
@haukepetersen, with the new gpio design we said, that it is up to the user to know his board and **not**  try to initialize "invalid" port/pin combinations. On the atmega2560 there are IIRC just some external interrupt lines. Do you think it makes sense to add a check here if the port/pin **is** valid?

haukepetersen replied:
Yes, definitely! I would suggest something like this:
```c
if (!((port == PORT_D) && (pin < 4)) || ((port == PORT_E) && (pin >= 4)))) {
    return -1;   // no EXTI possible
}
```

PeterKietzmann commented:
The interrupt external interrupt on `E5` seems broken. Can you please check that?!

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -0,0 +1,244 @@
+ISR(INT5_vect, ISR_BLOCK)
+{
+    irq_handler(_pin_num(GPIO(PORT_E, 5))); /**< predefined pin */
+}
```
Currently I'm thinking if we could make this a bit prettier. But as we don't define those vectors...

PeterKietzmann replied:
@N8Fear you wrote the [startup code](https://github.com/RIOT-OS/RIOT/blob/master/cpu/atmega2560/startup.c#L48) for that board. Do you have an idea how to write this code prettier and less redundant?

PeterKietzmann commented:
@ReneHerthel we're close before an ACK :-) . If the functional-issues are clarified I'm fine. The rest we can still do in follow-up PRs.

ReneHerthel commented:
Rebase.

ReneHerthel commented:
Squashed commits.

PeterKietzmann commented:
@ReneHerthel did you work on my last comments? If so, you should not have squashed them. Anyway, please give the commit message a reasonable name like "cpu/atmega2560: initial import of a gpio driver" or similar and don't use the word "squash" in that message (at least not when it is no commit that should be squashed)

ReneHerthel commented:
@PeterKietzmann Okey.
Yes I worked on it, but the gpio test doesn't work really... 
My board doesn't respond when I enter some of these shell commands or it reboots all the time, hmm when I write my own test, there is no problem. Did I forgot something?

PeterKietzmann commented:
Some comments above I mentioned that you have to adapt the standard baud rate for the UART communication to `9600` [here](https://github.com/RIOT-OS/RIOT/blob/master/boards/arduino-mega2560/include/board.h#L46). With that change, the example works (except the described issues). Please check what’s going wrong with the read function and the external interrupt on pin `E5`

ReneHerthel commented:
@PeterKietzmann Aaah, now it works! ;-) Thanks.
Okey, I'll look at the issues.

ReneHerthel commented:
@PeterKietzmann What exactly is broken with the E5?
When I try to initialize it and check if an interrupt appears, everything seems to be ok.
If I read the same pin after that, there is something like this:

```javascript
INFO # INT: external interrupt from pin 0
INFO # INT: external interrupt from pin 0
INFO # INT: external interrupt from pin 0
```

PeterKietzmann commented:
@ReneHerthel did you delete you last message? Anyway. 1. So you are able to read `LOW` and `high` values when any pin is initialized with `init_in`? 2. What happens if you initialize `E5` with `init_int` (as external interrupt pin) and then you apply the regarding signal to that pin? Last time it caused hard faults here. Now these faults are not visible and the program just hangs

ReneHerthel commented:
@PeterKietzmann 
Hmm I see, you mean that strange output..

INFO # INT: external interrupt from pin 0
INFO # INT: external interrupt from pin 0
INFO # INT: external interrupt from pin 0
INFO # INT: external interrupt from pin 0

or something else?
This appears only with E5, yes.
Hmm..

ReneHerthel commented:
@PeterKietzmann Did I ?
1. Yes, I can initialize a pin as input and 'set it' to HIGH or 'clear it' to LOW.
2. When I initialize E5 and try to read it or do something else with it, there are these 4 messages, like in my last coment.

PeterKietzmann commented:
Don't know if I understand you correctly or vice versa. Assuming you do the following:

- `init_in 0 0`
- apply a "HIGH" level to that pin
- `read 0 0` 

Do you get the "correct" output? That means the application prints "HIGH" when you applied a high value (e.g. 3,3V) and "LOW" when you apply `GND`? For me it is `LOW` all the time.

I understood that something with the external interrupt line on `E5` goes wrong. I thought as you are the most experienced guy in this situation, you may figure out what is wrong there. Any chance?

haukepetersen commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -0,0 +1,244 @@
+    irq_handler(_pin_num(GPIO(PORT_E, 4))); /**< predefined pin */
+}
+
+ISR(INT5_vect, ISR_BLOCK)
```
You are missing INT6 and INT7...

ReneHerthel commented:
@PeterKietzmann Hmm, yes it is always low, when you apply `3.3V` or `GND`.
It's only `HIGH`, when the bit is set in the register.
If you set the input to `HIGH` with `set 0 0`, the output after `read 0 0` is `HIGH`.
If you clear the same pin the output is `LOW`.

But you mean, it should print `HIGH`, when you attach a wire to the pin with `GND` or `3.3V`?
The gpio_read function is only looking at the bit, if it's set or not.

Or you think that the register changed automatically, when you apply something with an high value or GND?

And the E5 thing, no idea at the moment :-/

PeterKietzmann commented:
@ReneHerthel please adapt Haukes proposed changes. Regarding the read function: of course you want to get to logical value that corresponds to an external "logocal" signal. Why should you want to read a value that you set by yourself? Then the function is not implemented correctly. Regarding interrupts: Let's see if the same applies to INT6 and INT7. Maybe that gives exposure. If not we need to spend some time in this together (or so)..

ReneHerthel commented:
@PeterKietzmann Changes are adapted.
I think I got it.. we used before only PORTx and DDRx, now we need the PINx register to read the input.

PeterKietzmann commented:
@ReneHerthel honestly I don't know either. But as this is a CPU dependant implementation we should not care. Imagine this CPU is used on an other board design. There, it may appear that all interrupt lines have a hardware pin. So the `read` function works now as expected?  I'll test it next week and then the last thing should be the interrupt-bug.

ReneHerthel commented:
@PeterKietzmann Ah, okey.
I think `E5` is already fixed.
I can initialize `E5` as interrupt and connect it with another pin, so you can toggle the other pin to do some interrupts on `E5`.
`Read` should work, also on `E5`.
You can apply `3.3V` to read `HIGH` or `GND`to read `LOW` on an input pin.

Anyway, I'll come tomorrow, so we can finish this and maybe test it with an oscilloscope :-)

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright (C) 2015 haw-hamburg
```
Write it like "HAW Hamburg" please

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -0,0 +1,255 @@
+#define GPIO_OFFSET_PORT_H      (0xCB)
+#define GPIO_OFFSET_PIN_PORT    (0x02)
+#define GPIO_OFFSET_PIN_PIN     (0x03)
+#define GPIO_EXT_INT_NUMOF       7
```
Please add parentheses around the number

PeterKietzmann commented on boards/arduino-mega2560/include/periph_conf.h:
```diff
@@ -14,6 +14,7 @@
  * @brief       Peripheral MCU configuration for the Arduino Mega 2560 board
  *
  * @author      Hinnerk van Bruinehsen <h.v.bruinehsen@fu-berlin.de>
+ * @author      René Herthel <rene.herthel@haw-hamburg.de>
```
If you like you can add the HAW copyright under FU Berlin. **If** you do so, don't forget it is 2015 currently

PeterKietzmann commented on cpu/atmega2560/periph/gpio.c:
```diff
@@ -0,0 +1,255 @@
+int gpio_init_int(gpio_t pin, gpio_pp_t pullup, gpio_flank_t flank,
+                  gpio_cb_t cb, void *arg)
+{
+    uint8_t pin_num = _pin_num(pin);
```
IMO the last thing to do is to add a check whether the pin to initialize as interrupt source is valid (you know, there are jsut these some pins for external interrupts). I see two different possibilities, you should decide on your own which seems more efficient: 1. Compare [this implementation](https://github.com/RIOT-OS/RIOT/blob/master/cpu/samd21/periph/gpio.c#L40) 2. If it's just one or two lines you can simply check by `if`

PeterKietzmann commented:
@ReneHerthel, now everything works as expected and the code quality has **really** improved over the weeks. Please address my last minor comments and add a commit. I think we can merge it then :-)

PeterKietzmann commented:
Travis fails for two groups and I don't understand why. Could it be related to #3777?

```
$ source ./dist/tools/pr_check/check_labels.sh

$ test -z "$TRAVIS_PULL_REQUEST" || test "$BUILDTEST_MCU_GROUP" = "static-tests" || check_gh_label "Ready for CI build" || exit 1
```
@haukepetersen , @kaspar030, @phiros any ideas?

kaspar030 commented:
@PeterKietzmann #3777 is unlikely to cause this.

kaspar030 commented:
But travis seems to be very unresponsive right now... Can't even restart the build.

PeterKietzmann commented:
Okay. Maybe we should give him some time and try to restart later...

PeterKietzmann commented:
It was a long way. But finally things come to an end :-) . ACK and go

PeterKietzmann commented:
@ReneHerthel congratulations for your first contribution!

PeterKietzmann commented:
@OlegHahm and sorry for merging that in FeatureFreeze. It was important due to motivational reasons ;-)

LudwigKnuepfer commented:
@ReneHerthel please document the pin layout in the wiki
